### PR TITLE
feat(core): add OIDC SSO connector class

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

- Define a parent `OidcConnector` class. This class provides the basic functionality to connect with an OIDC provider. All the OIDC single sign-on connectors should extend this class. Integrated based on our existing social oidc-connector. 
- Define a `SingleSignOn` abstract class, that limits the necessary properties and methods for all the SSO connectors.
- Add `OidcSsoConnector` class.  Extend `OidcConnector`
- Define the connector factory type and the list of supported SSO connector factories. 

## Testing

UT cases added for the util methods.
Integration tests will be added along with later API PR. 

## Checklist

- [ ] `.changeset`
- [x] unit tests
- [ ] integration tests
- [ ] necessary TSDoc comments

### CODE REVIEW

File: packages/core/src/sso/types/index.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { type SsoConnector } from '@logto/schemas';
+import { type ZodType } from 'zod';
+
+import type TenantContext from '#src/tenants/TenantContext.js';
+
+export abstract class SsoClass {
```
Comment (gao-sun) [inline packages/core/src/sso/types/index.ts:6]:
```suggestion
export abstract class SingleSignOn {
```

File: packages/core/src/sso/OidcSsoConnector/index.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+import { fetchOidcConfig, fetchToken, getIdTokenClaims } from '#src/sso/utils/oidc.js';
+import type TenantContext from '#src/tenants/TenantContext.js';
+
+import { SsoClass, SsoProviderName } from '../types/index.js';
+import { type BaseOidcConfig } from '../types/oidc.js';
+
+import { configGuard, type ConfigType } from './config.js';
+
+class OidcSsoConnector extends SsoClass {
```
Comment (gao-sun) [inline packages/core/src/sso/OidcSsoConnector/index.ts:21]: might be better:

```ts
class OidcConnector {
  // ... OIDC stuff
}

abstract class SingleSignOn {
  // ... abstract stuff
}
```

finally

```ts
class OidcSingleSignOn extends OidcConnector implements SingleSignOn {}
```

**Review by wangsijie — APPROVED** [review]

File: packages/core/src/sso/OidcConnector/utils.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    return camelcaseKeys(result.data);
+  } catch (error: unknown) {
+    if (error instanceof HTTPError) {
+      throw new ConnectorError(ConnectorErrorCodes.General, error.response.body);
+    }
+    throw error;
+  }
+};
+
+const issuedAtTimeTolerance = 60;
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/core/src/sso/OidcConnector/utils.ts:80]: Follow our js SDK.

File: packages/core/src/sso/OidcConnector/utils.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  config: BaseOidcConfig,
+  nonceFromSession?: string
+) => {
+  try {
+    const { payload } = await jwtVerify(idToken, createRemoteJWKSet(new URL(config.jwksUri)), {
+      issuer: config.issuer,
+      audience: config.clientId,
+    });
+
+    if (Math.abs((payload.iat ?? 0) - Date.now() / 1000) > issuedAtTimeTolerance) {
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/core/src/sso/OidcConnector/utils.ts:93]: sure thing. It should be a standard token claim.

File: packages/core/src/sso/OidcSsoConnector/index.ts (hunk 2 above)
Comment (simeng-li) [reply packages/core/src/sso/OidcSsoConnector/index.ts]: updated

File: packages/core/src/sso/types/index.ts (hunk 1 above)
Comment (simeng-li) [reply packages/core/src/sso/types/index.ts]: updated

**Review by gao-sun — COMMENTED** [review]
this is actually too large. split into several PRs if possible

**Pull request merged**